tools/autograd/gen_variable_type.py

```diff
@@ -483,11 +483,29 @@ def is_differentiable(arg):
         if 'Tensor' not in arg['type']:
             return False
         if arg['dynamic_type'] in {'IndexTensor', 'BoolTensor'}:
+            # TODO: Enable this after native_functions.yaml schema unification.
+            # These are necessary for legacy code and should be
+            # used by legacy code only!
+            # assert name.startswith('_th_'), \
+            # "IndexTensor and BoolTensor are restricted to legacy _th_ functions only.
             return False
         return True
 
+    def find_args_with_derivatives(differentiable_inputs):
+        """Find arguments that have derivative definitions"""
+        if func is None:
+            return differentiable_inputs
+        names = set(name for d in func['derivatives'] for name in d['var_names'])
+        differentiable = [arg for arg in differentiable_inputs if arg['name'] in names]
+        if len(differentiable) != len(names):
+            missing = names - set(arg['name'] for arg in differentiable)
+            raise RuntimeError('Missing arguments for derivatives: {} in {}'.format(missing, func['name']))
+        return differentiable
+
     inputs = [arg for arg in arguments if not arg.get('output', False)]
     differentiable_inputs = list(filter(is_differentiable, inputs))
+    args_with_derivatives = find_args_with_derivatives(differentiable_inputs)
```

Comment:
I think this renaming is actually worse.  The previous naming IMO had 2 problems:
1) one thing was called gradients and one was called derivatives and it's not clear what difference was implied and/or why there wasn't just one concept.
2) both `args_with_derivatives` and `not_differentiable_args` are both called args, but for one it's dict of the entire arg and the other it's just a name.

This renaming doesn't fix 2) and it hides the difference in 1), i.e. there are still two concepts but now they have the same name.

Reply:
On 1) I don't think there was a difference between the two. The names were used interchangeably. 

On 2), i'll add "_names" to "not_differentiable_args".